sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -925,18 +925,20 @@ public boolean start() throws IOException {
       consumer = source.consumerFactoryFn.apply(source.consumerConfig);
       consumer.assign(source.assignedPartitions);
 
-      // seek to consumedOffset + 1 if it is set
       for (PartitionState p : partitionStates) {
-        if (p.consumedOffset >= 0) {
-          LOG.info("{}: resuming {} at {}", name, p.topicPartition, p.consumedOffset + 1);
+        if (p.consumedOffset >=  0) {
           consumer.seek(p.topicPartition, p.consumedOffset + 1);
         } else {
-          LOG.info("{}: resuming {} at default offset", name, p.topicPartition);
+          // set consumed offset to (next offset - 1), otherwise checkpoint would contain invalid
+          // offset until we read first record from this partition.
+          p.consumedOffset = consumer.position(p.topicPartition) - 1;
```

Comment:
**Edge case**, but still: If this is a new/empty partition, first micro-batch read will be skipped and the latest (was actually `0`) will advance.
Maybe use `null` or `-2` for a "never tried to read yet" flag. Or add one.